src/distributed/topology_leader.ts

```diff
@@ -203,53 +207,77 @@ export class TopologyLeader {
                     if (!perform_loop || alive_workers.length == 0) {
                         return xcallback();
                     }
-                    self.storage.getTopologyStatus((err, topologies) => {
+                    self.storage.getTopologyStatus((err, topologies_all) => {
                         if (err) return xcallback(err);
-                        topologies = topologies.filter(x => x.enabled);
-                        topologies.forEach(x => {
-                            x.weight = x.weight || 1;
-                            x.worker_affinity = x.worker_affinity || [];
-                            if (x.status == "" || x.status == intf.Consts.TopologyStatus.unassigned) {
-                                for (let worker of alive_workers) {
-                                    let name = worker.name;
-                                    if (name == x.worker) {
-                                        let old_weight = 0;
-                                        if (worker_weights.has(name)) {
-                                            old_weight = worker_weights.get(name);
-                                        }
-                                        worker_weights.set(name, old_weight + x.weight);
-                                        break;
+                        topologies_disabled = topologies_all.filter(x => !x.enabled);
+                        topologies_enabled = topologies_all.filter(x => x.enabled);
+                        xcallback();
+                    });
+                },
+                (xcallback) => {
+                    if (!perform_loop || alive_workers.length == 0) {
+                        return xcallback();
+                    }
+                    let targets = topologies_enabled
+                        .filter(x => x.status == intf.Consts.TopologyStatus.running && x.worker == null);
+                    async.each(
+                        targets,
+                        (item: intf.TopologyStatus, xxcallback) => {
+                            // strange, topology marked as enabled and running, but no worker specified
+                            // mark it as unassinged.                        
+                            log.logger().important(this.log_prefix + "Topology marked as running and enabled, but no worker specified: " + item.uuid);
+                            self.storage.setTopologyStatus(item.uuid, intf.Consts.TopologyStatus.unassigned, null, xxcallback);
+                        },
+                        xcallback
+                    );
+                },
+                (xcallback) => {
```

Comment:
some comments on various stages would be nice, it's hard to follow what's being done in each callback

Reply:
Working on it.